cardano-crypto-praos/src/Cardano/Crypto/VRF/Praos.hs

```diff
@@ -0,0 +1,464 @@
+newtype SK = SK { unSK :: ForeignPtr SKValue }
+  deriving (Generic)
+
+-- | Public key.
+newtype PK = PK { unPK :: ForeignPtr PKValue }
+  deriving (Generic)
+
+-- | A proof, as constructed by the 'prove' function.
+newtype Proof = Proof { unProof :: ForeignPtr ProofValue }
+  deriving (Generic)
+
+-- | Hashed output of a proof verification, as returned by the 'verify'
+-- function.
+newtype Output = Output { unOutput :: ForeignPtr OutputValue }
+  deriving (Generic)
+
+-- Raw low-level FFI bindings.
+--
+foreign import ccall "crypto_vrf_proofbytes" crypto_vrf_proofbytes :: CSize
+foreign import ccall "crypto_vrf_publickeybytes" crypto_vrf_publickeybytes :: CSize
+foreign import ccall "crypto_vrf_secretkeybytes" crypto_vrf_secretkeybytes :: CSize
+foreign import ccall "crypto_vrf_seedbytes" crypto_vrf_seedbytes :: CSize
+foreign import ccall "crypto_vrf_outputbytes" crypto_vrf_outputbytes :: CSize
+
+foreign import ccall "crypto_vrf_keypair_from_seed" crypto_vrf_keypair_from_seed :: PKPtr -> SKPtr -> SeedPtr -> IO CInt
+foreign import ccall "crypto_vrf_sk_to_pk" crypto_vrf_sk_to_pk :: PKPtr -> SKPtr -> IO CInt
+foreign import ccall "crypto_vrf_sk_to_seed" crypto_vrf_sk_to_seed :: SeedPtr -> SKPtr -> IO CInt
+foreign import ccall "crypto_vrf_prove" crypto_vrf_prove :: ProofPtr -> SKPtr -> Ptr CChar -> CULLong -> IO CInt
+foreign import ccall "crypto_vrf_verify" crypto_vrf_verify :: OutputPtr -> PKPtr -> ProofPtr -> Ptr CChar -> CULLong -> IO CInt
+
+foreign import ccall "crypto_vrf_proof_to_hash" crypto_vrf_proof_to_hash :: OutputPtr -> ProofPtr -> IO CInt
+
+foreign import ccall "randombytes_buf" randombytes_buf :: Ptr a -> CSize -> IO ()
+
+-- Key size constants
+
+certSizeVRF :: Int
+certSizeVRF = fromIntegral crypto_vrf_proofbytes
+
+signKeySizeVRF :: Int
+signKeySizeVRF = fromIntegral crypto_vrf_secretkeybytes
+
+verKeySizeVRF :: Int
+verKeySizeVRF = fromIntegral crypto_vrf_publickeybytes
+
+vrfKeySizeVRF :: Int
+vrfKeySizeVRF = fromIntegral crypto_vrf_outputbytes
+
+-- | Allocate a 'Seed' and attach a finalizer. The allocated memory will not be initialized.
+mkSeed :: IO Seed
+mkSeed = do
+  ptr <- mallocBytes (fromIntegral crypto_vrf_seedbytes)
+  Seed <$> newForeignPtr finalizerFree ptr
+
+-- | Generate a random seed.
+-- Uses 'randombytes_buf' to create random data.
+genSeed :: IO Seed
+genSeed = do
+  seed <- mkSeed
+  withForeignPtr (unSeed seed) $ \ptr ->
+    randombytes_buf ptr crypto_vrf_seedbytes
+  return seed
+
+seedFromBytes :: ByteString -> Seed
+seedFromBytes bs | BS.length bs < fromIntegral crypto_vrf_seedbytes =
+  error "Not enough bytes for seed"
+seedFromBytes bs = unsafePerformIO $ do
+  seed <- mkSeed
+  withForeignPtr (unSeed seed) $ \ptr ->
+    BS.useAsCString bs $ \cstr ->
+      copyBytes (castPtr ptr) cstr (fromIntegral crypto_vrf_seedbytes)
+  return seed
+
+-- | Convert an opaque 'Seed' into a 'ByteString' that we can inspect. Note
+-- that this will leak the seed into unprotected memory.
```

Comment:
Although technically we're not allocating the seed in protected memory anyway...

Reply:
Right, this was from when we were operating under the assumption that we would in fact mlock the seed.